### PR TITLE
document additional output options

### LINKED ISSUE #914: Enable, test and document the Kafka output
It would be great to investigate if the APM Server can support Kafka as output.

- [x] test that enabling the Kafka output works as expected
- [ ] make sure we have a test in CI that tests APM Server with the Kafka output
- [ ] be aware of sourcemap support when using Kafka
- [ ] document Kafka usage with APM Server

### PR DESCRIPTION
**Pull request by graphaelli**
In addition to enabling previously inherited-from-beats but suppressed output options, these changes:

* document `fields` configuration
* update docs previously synced from beats
* reconfigure securing apm server

closes elastic/apm-server#914
closes elastic/apm-server#928

### CODE REVIEW

File: docs/configuring-output.asciidoc [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+== Output Configuration
+
+An output direct data from APM Server to the desired storage system.
+A typical deployment uses Elasticsearch as the output.
+Options like Logstash and Kafka are popular with users requiring additional data enrichment and highly scalable setups.
+
+[float]
+=== Sourcemaps
+
+Sourcemaps can be uploaded through various outputs but can currently only be applied when Elasticsearch is the output.
```
Comment (jalvz) [inline docs/configuring-output.asciidoc]: If you upload a sourcemap to kafka, and then change the output to elasticsearch, the sourcemap can't be applied, right?
So i'd say they can only be uploaded and applied to Elasticsearch.

File: docs/copied-from-beats/shared-config-ingest.asciidoc [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+//// this file. When using cross references, make sure that the cross
+//// references resolve correctly for any files that include this one.
+//// Use the appropriate variables defined in the index.asciidoc file to
+//// resolve Beat names: beatname_uc and beatname_lc.
+//// Use the following include to pull this content into a doc file:
+//// include::../../libbeat/docs/shared-config-ingest.asciidoc[]
+//////////////////////////////////////////////////////////////////////////
+
+[[configuring-ingest-node]]
+== Parse data by using ingest node
```
Comment (jalvz) [inline docs/copied-from-beats/shared-config-ingest.asciidoc]: did we test this with apm server?

File: docs/configuring-output.asciidoc [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+[source,yaml]
+------------------------------------------------------------------------------
+fields: {project: "myproject", instance-id: "574734885120952459"}
+------------------------------------------------------------------------------
+
+To store the custom fields as top-level fields, set the `fields_under_root` option to true.
+This is not recommended as backwards compatibility can not be ensured as new fields are added to APM documents.
+
+include::./copied-from-beats/outputconfig.asciidoc[]
```
Comment (simitt) [inline docs/configuring-output.asciidoc]: Since this included file is on the same level as `output-config` it creates a separate html page, leading to 

screen shot 2018-07-10 at 10 46 57

I suggest to add a check in the libbeat file if there is an additional config file like e.g. `beats-specific-output-config` and include it if defined. I quickly tried it out and it would give you something like this:

screen shot 2018-07-10 at 11 02 36

by adding:

* index.asciidoc:
`:beats-specific-output-config: ./../configuring-output.asciidoc`
* docs/copied-from-beats/outputconfig.asciidoc:
```
ifdef::beats-specific-output-config[]
  include::{beats-specific-output-config}[]
endif::[]
```
* docs/configuring.asciidoc
```
-include::./configuring-output.asciidoc[]
+include::./copied-from-beats/outputconfig.asciidoc[]
+include::./copied-from-beats/shared-config-ingest.asciidoc[]
```
* docs/configuring-output.asciidoc
```
-== Output Configuration

-== `fields`
+=== `fields`

-include::./copied-from-beats/outputconfig.asciidoc[]
-include::./copied-from-beats/shared-config-ingest.asciidoc[]
```

File: docs/copied-from-beats/outputconfig.asciidoc [hunk 4]
```diff
@@ -32,6 +32,9 @@ The following topics describe how to configure each supported output:
 * <<file-output>>
 * <<console-output>>
 
+If you've secured the {stack}, also read <<securing-{beatname_lc}>> for more about
```
Comment (simitt) [inline docs/copied-from-beats/outputconfig.asciidoc]: You probably need an additional check wheter or not `redis-output` is defined, as we have not tested nor documented it in the `apm-server.yml`.

File: docs/copied-from-beats/template-config.asciidoc [hunk 5]
```diff
@@ -87,4 +87,16 @@ ifeval::["{beatname_lc}"!="apm-server"]
 
 NOTE: With append_fields only new fields can be added an no existing one overwritten or changed. This is especially useful if data is collected through the http/json metricset where the data structure is not known in advance. Changing the config of append_fields means the template has to be overwritten and only applies to new indices. If there are 2 Beats with different append_fields configs the last one writing the template will win. Any changes will also have an affect on the Kibana Index pattern.
 
+*`setup.template.json.enabled`*:: Set to true to load a json based template file. Specify the path to your Elasticsearch
```
Comment (jalvz) [inline docs/copied-from-beats/template-config.asciidoc]: this has some implications for the apm server, right? (like breaking the UI)

File: docs/common-problems.asciidoc [hunk 6]
```diff
@@ -55,3 +55,67 @@ you can try to:
 * <<adjust-concurrent-requests>>
 * <<add-apm-server-nodes>>
 
+[float]
+[[ssl-client-fails]]
+=== SSL client fails to connect to Logstash?
```
Comment (simitt) [inline docs/common-problems.asciidoc]: Is all of this really logstash specific? Can't you also run into SSL issues with ES, Kibana, Kafka?

File: _meta/beat.yml [hunk 7]
```diff
@@ -414,6 +414,241 @@ output.elasticsearch:
   # Pretty print json event
   #pretty: false
 
+#----------------------------- Logstash output ---------------------------------
+#output.logstash:
+  # Boolean flag to enable or disable the output module.
+  #enabled: true
```
Comment (jalvz) [inline _meta/beat.yml]: this is not enabled by default, right? (same for kafka bellow)

File: docs/copied-from-beats/outputconfig.asciidoc [hunk 8]
```diff
@@ -833,15 +897,12 @@ The Redis port to use if `hosts` does not contain a port number. The default is
...
-
-The name of the Redis list or channel the events are published to. The default is
-"{beatname_lc}".
+The index name added to the events metadata for use by Logstash. The default is "{beatname_lc}".
 
 ===== `key`
 
-The name of the Redis list or channel the events are published to. The default is
-"{beatname_lc}".
+The name of the Redis list or channel the events are published to. If not
```
Comment (jalvz) [inline docs/copied-from-beats/outputconfig.asciidoc]: are we officially supporting Redis too?

File: docs/configuring-output.asciidoc (hunk 1 above)
Comment (graphaelli) [reply docs/configuring-output.asciidoc]: If you upload via kafka and then change the output to elasticsarch the sourcemap will be applied.  If you leave the output as kafka, the sourcemap can not be retrieved.  Thanks for opening #1104 to discuss this.

File: _meta/beat.yml (hunk 7 above)
Comment (graphaelli) [reply _meta/beat.yml]: good point, fixed

File: docs/copied-from-beats/shared-config-ingest.asciidoc (hunk 2 above)
Comment (graphaelli) [reply docs/copied-from-beats/shared-config-ingest.asciidoc]: Yes, in #1032

File: docs/common-problems.asciidoc (hunk 6 above)
Comment (graphaelli) [reply docs/common-problems.asciidoc]: Agreed, this was copied over from beats but now that you point it out it is odd.  removed

File: docs/security.asciidoc [hunk 9]
```diff
@@ -17,20 +17,4 @@ Both the agents and the APM servers have to be configured with the same secret t
 
 NOTE: The usage of a secret token only provides any security when used in combination with having SSL/TLS configured.
 
-[[ssl-setup]]
-[float]
-==== SSL/TLS setup
```
[the remark this answers is not in the document]
Comment (graphaelli) [reply docs/security.asciidoc]: That was a mistake, thanks!

File: docs/copied-from-beats/outputconfig.asciidoc (hunk 8 above)
Comment (graphaelli) [reply docs/copied-from-beats/outputconfig.asciidoc]: after discussion, no.  removed.

File: docs/configuring-output.asciidoc (hunk 3 above)
Comment (graphaelli) [reply docs/configuring-output.asciidoc]: thanks, this is a great idea

File: docs/copied-from-beats/template-config.asciidoc (hunk 5 above)
Comment (graphaelli) [reply docs/copied-from-beats/template-config.asciidoc]: yeah, it's probably not worth trying to explain that either, removed.

File: _meta/beat.yml [hunk 10]
```diff
@@ -414,6 +414,241 @@ output.elasticsearch:
...
+  #loadbalance: false
+
+  # Number of batches to be sent asynchronously to Logstash while processing
+  # new batches.
+  #pipelining: 2
+
+  # If enabled only a subset of events in a batch of events is transferred per
+  # transaction.  The number of events to be sent increases up to `bulk_max_size`
+  # if no error is encountered.
+  #slow_start: false
```
[the remark this answers is not in the document]
Comment (graphaelli) [reply _meta/beat.yml]: changed to `group`

File: docs/security.asciidoc [hunk 11]
```diff
@@ -17,20 +17,4 @@ Both the agents and the APM servers have to be configured with the same secret t
...
-
-To enable SSL/TLS you need a private key and a certificate issued by a certification authority (CA).
-Then you can specify the path to those files in the configuration properties
-`apm-server.ssl.key` and
-`apm-server.ssl.certificate`
-respectively.
-This will make the APM Server to serve HTTPS requests instead of HTTP.
-Hence, you also need to enable SSL in the agent.
```
[the remark this answers is not in the document]
Comment (graphaelli) [reply docs/security.asciidoc]: restored

File: docs/copied-from-beats/https.asciidoc [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+//// resolve Beat names: beatname_uc and beatname_lc.
+//// Use the following include to pull this content into a doc file:
+//// include::../../libbeat/docs/https.asciidoc[]
+//// This content is structured to be included as a whole file.
+//////////////////////////////////////////////////////////////////////////
+
+To secure the communication between {beatname_uc} and Elasticsearch, you can use
+HTTPS and basic authentication. Basic authentication for Elasticsearch is
+available when you enable {security} (see
+{securitydoc}/xpack-security.html[Securing the {stack}] and <<securing-beats>>).
```
[the remark this answers is not in the document]
Comment (graphaelli) [reply docs/copied-from-beats/https.asciidoc]: nice catch, fixed

**Comment by graphaelli** [comment]
Thanks for the great feedback.  The index looks like this now:

image

This will depend on some update beats to support adding beat-specific text.  I'll open a pull request with those changes once this pull request takes shape.

File: docs/copied-from-beats/command-reference.asciidoc [hunk 13]
```diff
@@ -467,7 +453,10 @@ environment without actually running {beatname_uc} and ingesting data.
 *FLAGS*
 
 *`--dashboards`*::
-Sets up the Kibana dashboards only.
+Sets up the Kibana dashboards only. This option loads the dashboards from the
+{beatname_uc} package. For more options, such as loading customized dashboards,
+see {beatsdevguide}/import-dashboards.html[Importing Existing Beat Dashboards]
+in the _Beats Developer Guide_.
```
Comment (jalvz) [inline docs/copied-from-beats/command-reference.asciidoc]: is this relevant to us?

File: docs/copied-from-beats/dashboards.asciidoc [hunk 14]
```diff
@@ -18,14 +18,21 @@ command (as described here) or
...
 Make sure Kibana is running before you perform this step. If you are accessing a
 secured Kibana instance, make sure you've configured credentials as described in
 <<{beatname_lc}-configuration>>.
 
 To set up the Kibana dashboards for {beatname_uc}, use the appropriate command
-for your system.
+for your system. The command shown here loads the dashboards from the {beatname_uc}
+package. For more options, such as loading customized dashboards, see
+{beatsdevguide}/import-dashboards.html[Importing Existing Beat Dashboards] in
+the _Beats Developer Guide_.
```
Comment (jalvz) [inline docs/copied-from-beats/dashboards.asciidoc]: duplicated from the command reference, is intentional?

File: docs/copied-from-beats/security/securing-beats.asciidoc [hunk 15]
```diff
@@ -1,33 +1,79 @@
...
+{xpack-ref}/xpack-security.html[Securing the {stack}].
+
+[float]
+=== {beatname_uc} features that require authorization
+
+After securing {beatname_uc}, make sure your users have the roles (or associated
+privileges) required to use these {beatname_uc} features. You must create the
+++{beat_default_index_prefix}_writer++ and
+++{beat_default_index_prefix}_reader++ roles (see <<beats-basic-auth>> and
+<<beats-user-access>>). The `machine_learning_admin` and `kibana_user` roles are
```
Comment (jalvz) [inline docs/copied-from-beats/security/securing-beats.asciidoc]: does this show up or is behind a conditional somewhere? 
if it does: is the machine learning role relevant?

File: docs/copied-from-beats/shared-config-ingest.asciidoc (hunk 2 above)
Comment (jalvz) [reply docs/copied-from-beats/shared-config-ingest.asciidoc]: nice example you got there! totally missed it

File: docs/copied-from-beats/shared-securing-beat.asciidoc [hunk 16]
```diff
@@ -0,0 +1,38 @@
+[id="securing-{beatname_lc}"]
```
Comment (jalvz) [inline docs/copied-from-beats/shared-securing-beat.asciidoc]: Since security is an advanced topic and not the thing that you typically try first, I'd move the whole "Securing APM Server" section towards the end; maybe just before the exported fields.

File: docs/configuring.asciidoc [hunk 17]
```diff
@@ -18,10 +18,11 @@ The following topics describe how to configure APM Server:
 * <<configuration-path>>
 --
 
-:only-elasticsearch:
```
Comment (jalvz) [inline docs/configuring.asciidoc]: I guess this is not easy, but would be nice to mention that the APM UI only works with elasticsearch output...

File: docs/copied-from-beats/security/linux-seccomp.asciidoc [hunk 18]
```diff
@@ -0,0 +1,93 @@
+[[linux-seccomp]]
```
[the remark this answers is not in the document]
Comment (graphaelli) [reply docs/copied-from-beats/security/linux-seccomp.asciidoc]: #928 has been open for that, I can pick it up before merging this though I don't expect any issues.

[the remark this answers is not in the document]
Comment (graphaelli) [reply docs/copied-from-beats/security/linux-seccomp.asciidoc]: This has been enabled by default since it was merged - libbeat already logs installation status on startup - 

```
{
  "level":"info",
  "logger":"seccomp",
  "caller":"seccomp/seccomp.go:116",
  "message":"Syscall filter successfully installed"
}
```

I tested enable/disable and the `kill_process ` action - nothing was triggered when hitting the various apm-server endpoints, as expected.  #928 can be closed once this pull request is merged.

File: docs/copied-from-beats/dashboards.asciidoc (hunk 14 above)
Comment (graphaelli) [reply docs/copied-from-beats/dashboards.asciidoc]: Since this is copied directly from beats I can only guess this is because we don't expect users to read all docs end to end - a little repetition seems ok to me.  I've removed this and the section in command-reference as I agree they are not totally relevant to apm-server.

File: docs/copied-from-beats/command-reference.asciidoc (hunk 13 above)
Comment (graphaelli) [reply docs/copied-from-beats/command-reference.asciidoc]: good call, removed.

File: docs/copied-from-beats/shared-securing-beat.asciidoc (hunk 16 above)
Comment (graphaelli) [reply docs/copied-from-beats/shared-securing-beat.asciidoc]: Agreed it makes sense a bit later, once you have the basics of configuration set - moved it down a bit.

File: docs/copied-from-beats/security/securing-beats.asciidoc (hunk 15 above)
Comment (graphaelli) [reply docs/copied-from-beats/security/securing-beats.asciidoc]: It is but it's broken, fixing.

File: docs/configuring.asciidoc (hunk 17 above)
Comment (graphaelli) [reply docs/configuring.asciidoc]: In `Load the Elasticsearch index template` we will have a `load the template manually` section for users of non-elasticsearch outputs.

File: docs/copied-from-beats/security/linux-seccomp.asciidoc (hunk 18 above)
[the remark this answers is not in the document]
Comment (jalvz) [reply docs/copied-from-beats/security/linux-seccomp.asciidoc]: great, thanks for digging in  👍
